src/board.rs

```diff
@@ -80,12 +80,38 @@ impl Board {
         }
     }
 
-    pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult<()> {
+    pub fn move_piece<F>(
+        &mut self,
+        from: Position,
+        to: Position,
+        promotion_callback: F,
+    ) -> ChessResult<()>
+    where
+        F: Fn(Player) -> Piece,
+    {
         self.is_move_valid(from, to)?;
-        self.squares[to.x][to.y] = self.get_piece(from).take();
+
+        if let Some(Piece::Pawn(player)) = self.get_piece(from).take() {
+            if (player == Player::White && to.y == 7) || (player == Player::Black && to.y == 0) {
+                let promoted_piece = promotion_callback(player);
+                self.promote_pawn(to, promoted_piece)?;
+            }
+        }
         Ok(())
     }
 
+    pub fn promote_pawn(&mut self, position: Position, new_piece: Piece) -> ChessResult<()> {
+        // The piece should be a pawn and should be on a promotion square
+        if let Some(Piece::Pawn(player)) = self.get_piece(position) {
```

Comment:
So I wonder if we shouldn't be doing these checks here for performance. Maybe if we are going to use bitboard stuff for the bot anyway it's fine.